[1.21] Add support for the `c:register` channel registration payload

Shadows-of-Fire opened the pull request:
This PR adds support for the `c:register` payload, which allows declaring the play-phase channels during the configuration phase, as established by Fabric in https://github.com/FabricMC/fabric/pull/3244

To permit compatibility with fabric, the `c:version` payload is also here, though it is technically unsupported from our end. Fabric requires it be sent before sending `c:register`.  I don't _really_ see this version mechanism being useful atm and am aiming for a minimal implementation.

Aside from the explicit `c:register` support, this PR has a couple minor cleanup items:
1. Fixing the error messages in the exceptions thrown by `checkPacket` - some format strings are incorrectly using `{}` instead of `%s`.
2. Moves the channel attribute handling code to a new standalone class.
3. Fixes a couple style issues in `NetworkRegistry`.  At some point I should probably make braces on conditional statements required.

Closes #1051

Shadows-of-Fire committed:
Add support for c:register

Also does some misc cleanup to the network registry code, fixing the error messages from hasChannel and moving the attribute-handling code to a standalone class.

sciwhiz12 commented on patches/net/minecraft/client/multiplayer/ClientCommonPacketListenerImpl.java.patch:
```diff
@@ -35,6 +35,16 @@
 +        }
 +
++        if (p_295727_.payload() instanceof net.neoforged.neoforge.network.payload.CommonVersionPayload) {
++            // no-op
```
Explain in this comment why the `c:version` payload is a no-op. (Alternatively, explain it in `CommonVersionPayload` and reference that from this comment.)

sciwhiz12 commented on patches/net/minecraft/server/network/ServerCommonPacketListenerImpl.java.patch:
```diff
@@ -35,6 +35,16 @@
 +        }
 +
++        if (p_294276_.payload() instanceof net.neoforged.neoforge.network.payload.CommonVersionPayload) {
++            // no-op
```
Explain in this comment why the `c:version` payload is a no-op. (Alternatively, explain it in `CommonVersionPayload` and reference that from this comment.)

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/network/payload/CommonRegisterPayload.java:
```diff
@@ -0,0 +1,49 @@
+            ByteBufCodecs.VAR_INT, CommonRegisterPayload::version,
+            ByteBufCodecs.STRING_UTF8.map(CommonRegisterPayload::protocolById, ConnectionProtocol::id), CommonRegisterPayload::protocol,
+            ByteBufCodecs.collection(HashSet::new, ResourceLocation.STREAM_CODEC), CommonRegisterPayload::channels,
+            CommonRegisterPayload::new);
+    @Override
```
Please add a blank line between the field and the method.

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/network/payload/CommonRegisterPayload.java:
```diff
@@ -0,0 +1,49 @@
+ * @param protocol The {@link ConnectionProtocol} of the contained channels. One of "play" or "configuration". Currently NeoForge only expects and sends "play" channels via this payload.
+ * @param channels A list of all named channels available for the declared phase.
+ */
+public record CommonRegisterPayload(int version, ConnectionProtocol protocol, Set<ResourceLocation> channels) implements CustomPacketPayload {
```
Mark this class as `@ApiStatus.Internal`.

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/network/payload/CommonVersionPayload.java:
```diff
@@ -0,0 +1,35 @@
+ * Unsupported common version payload. Must be sent to fabric to support their implementation of c:register.
+ * Only legal version is 1.
+ */
+public record CommonVersionPayload(List<Integer> versions) implements CustomPacketPayload {
```
Mark this class as `@ApiStatus.Internal`.

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -78,17 +78,11 @@
-    private static final AttributeKey<NetworkPayloadSetup> ATTRIBUTE_PAYLOAD_SETUP = AttributeKey.valueOf("neoforge:payload_setup");
-    private static final AttributeKey<Set<ResourceLocation>> ATTRIBUTE_ADHOC_CHANNELS = AttributeKey.valueOf("neoforge:adhoc_channels");
-    private static final AttributeKey<ConnectionType> ATTRIBUTE_CONNECTION_TYPE = AttributeKey.valueOf("neoforge:connection_type");
-    private static final AttributeKey<PacketFlow> ATTRIBUTE_FLOW = AttributeKey.valueOf("neoforge:flow");
```
Why did you remove this channel attribute (and not mention it in the PR description as part of your cleanups)?

Shadows-of-Fire replied:
Forgot to mention it.  This attribute was unused, so there was no point to keep it while refactoring.

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/network/registration/ChannelAttributes.java:
```diff
@@ -0,0 +1,95 @@
+    /**
+     * Known common channels from {@link CommonRegisterPayload}.
+     */
+    public static final AttributeKey<Map<ConnectionProtocol, Set<ResourceLocation>>> COMMON_CHANNELS = AttributeKey.valueOf("neoforge:common_channels");
```
Why not a `SetMultimap` from Guava?

Shadows-of-Fire replied:
Don't know, but my source attachment is busted for guava in neodev atm, so I shall be leaving it as-is.

sciwhiz12 replied:
Perhaps then, we should add a comment to consider moving it to SetMultimap in the future?

Shadows-of-Fire replied:
Eh, probably not necessary.  If someone wants to move it they can do so with or without the todo.

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -597,20 +579,24 @@ public static boolean hasChannel(ICommonPacketListener listener, ResourceLocatio
+            // Otherwise check against all protocols.
+            if (protocol != null && payloadSetup.getChannels(protocol).containsKey(payloadId)) {
+                return true;
+            } else if (protocol == null && payloadSetup.channels().values().stream().anyMatch(map -> map.containsKey(payloadId))) {
```
Just a small fun comment: it took me looking at the code to realize why this looked strange to me -- `NetworkPayloadSetup` contains a map whose values are also maps. 😆

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -622,7 +608,7 @@ public static boolean hasChannel(Connection connection, @Nullable ConnectionProt
      */
     public static <T extends PacketListener> List<Packet<?>> filterGameBundlePackets(ChannelHandlerContext context, Iterable<Packet<? super T>> packets) {
-        final NetworkPayloadSetup payloadSetup = context.channel().attr(ATTRIBUTE_PAYLOAD_SETUP).get();
+        NetworkPayloadSetup payloadSetup = context.channel().attr(ChannelAttributes.PAYLOAD_SETUP).get();
```
Why not use the `ChannelAttributes#getPayloadSetup` method?

Shadows-of-Fire replied:
Because those helpers work on `Connection` (to avoid unwrapping the `Channel` at most call sites), but this call site is unique in that we don't have a `Channel`, we have a `ChannelHandlerContext`.

And while true they both have a method with signature `public Channel channel()`, there isn't a common interface to retrieve it by.

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -719,6 +704,39 @@ public static Set<ResourceLocation> getInitialServerUnregisterChannels() {
     }
 
+    /**
+     * Replaces any existing common channels with the incoming ones from a {@link CommonRegisterPayload}.
```
Is that the specified behavior of `c:register` on Fabric -- it _replaces_ all registrations with the registrations specified, instead of adding the new registrations? I would think it mimics `minecraft:register`, but I'd like to be sure just in case.

(@Technici4n, could you please advise on this?)

Shadows-of-Fire replied:
I wasn't sure, and didn't see it formally specified anywhere.  Having the behavior of `minecraft:register` would be weird because there isn't a `c:unregister` payload to complement that behavior.

sciwhiz12 replied:
We could check the implementation of the packet in Fabric API (and/or consult with @fabricmc maintainers) to see if it does a replacement or a simple addition.

Minecraftschurli replied:
> ... there isn't a c:unregister payload ...

there is a `minecraft:unregister` too

Shadows-of-Fire replied:
The `minecraft:` payloads do not interact with the channels supplied by the `c:` payload.  Thus I think the best implementation we can have is for `c:register` to replace the existing channels when it comes in, with the expectation that it be sent one time during configuration.

sciwhiz12 reviewed (changes requested):
Not tested, just a code review.

Technici4n commented:
We had some discussion on Discord a few days ago. It's not clear that we need to support this. An alternative would be to allow sending unknown packets for `OTHER` connections, at least in the play phase.

Shadows-of-Fire commented:
Just allowing any payload to be sent doesn't feel like the right solution, especially when this protocol exists and can be supported.  I don't want to have a situation where the contract of `hasChannel` is "check this, but the result might be invalid on an `OTHER` connection".

Technici4n commented:
> I don't want to have a situation where the contract of hasChannel is "check this, but the result might be invalid on an OTHER connection".

You can perform the check during the configuration phase if you want to support `OTHER` connections reliably.

Shadows-of-Fire commented:
Doing checks in the config phase only allows you to check for config channels, which isn't useful here.  There are only three effective solutions that might solve this problem:
1. Support `c:register` and allow the ecosystem to support it broadly so that channels are declared early in the config phase. (this solution)
2. Remove validation in `hasChannel`. Has the aforementioned side effect of creating a poor contract for the method, making code that has to interact with it awkward (the method would basically be invalid on `OTHER` connections, meaning consumers would need to always query the connection type _or_ hasChannel).
3. Provide an event that fires when `minecraft:register` is received (and channels are updated) so that consumers can fire login payloads as soon as the connection declares they're available.

Shadows-of-Fire committed:
Properly support c:version

Shadows-of-Fire commented:
Updated to address Sci's review comments - I just implemented the correct payload handler for `c:version`, though given there's only one version it's mostly superfluous.

sciwhiz12 committed:
Merge branch '1.21.x' into net-register

sciwhiz12 reviewed (approved):
Looks fine to me, I believe.

@Technici4n, please review or leave a comment if you will not be leaving a review, thanks.

Technici4n commented:
I'll have a look on Sunday latest.
